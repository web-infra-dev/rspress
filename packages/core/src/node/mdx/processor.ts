import path from 'node:path';
import { createProcessor } from '@mdx-js/mdx';
import type { Header, UserConfig } from '@rspress/shared';
import { extractTextAndId, loadFrontMatter } from '@rspress/shared/node-utils';

import type { PluginDriver } from '../PluginDriver';
import type { RouteService } from '../route/RouteService';
import {
  applyReplaceRules,
  escapeMarkdownHeadingIds,
  normalizePath,
} from '../utils';
import { createMDXOptions } from './options';
import type { TocItem } from './remarkPlugins/toc';
import type { PageMeta } from './types';

interface CompileOptions {
  source: string;
  filepath: string;

  // assume that the below instances are singleton, it will not change.
  docDirectory: string;
  config: UserConfig | null;
  routeService: RouteService | null;
  pluginDriver: PluginDriver | null;
}

async function compile(options: CompileOptions): Promise<string> {
  const { source, filepath, docDirectory, config, routeService, pluginDriver } =
    options;

  const mdxOptions = await createMDXOptions({
    config,
    docDirectory,
    filepath,
    pluginDriver,
    routeService,
  });
  // Separate frontmatter and content in MDX source
  const { frontmatter, emptyLinesSource } = loadFrontMatter(
    source,
    filepath,
    docDirectory,
    true,
  );

  // For loader error stack, so we use the source with frontmatter @see https://github.com/web-infra-dev/rspress/issues/2010
  const { replaceRules } = config ?? {};
  const replacedSource = applyReplaceRules(emptyLinesSource, replaceRules);

  // Support custom id like `#hello world {#custom-id}`.
  // TODO > issue: `#hello world {#custom-id}` will also be replaced.
  const preprocessedContent = escapeMarkdownHeadingIds(replacedSource);

  try {
    let pageMeta: PageMeta = {
      title: '',
      toc: [] as TocItem[],
      headingTitle: '',
    };

    const frontmatterTitle = extractTextAndId(frontmatter.title)[0];

    const compiler = createProcessor(mdxOptions);

    compiler.data('pageMeta' as any, { toc: [], title: '' });
    const vFile = await compiler.process({
      value: preprocessedContent,
      path: filepath,
    });

    const compileResult = String(vFile);
    const compilationMeta = compiler.data('pageMeta' as any) as {
      toc: Header[];
      title: string;
    };

    const headingTitle = extractTextAndId(compilationMeta.title)[0];
    pageMeta = {
      ...compilationMeta,
      title: frontmatterTitle || headingTitle,
      headingTitle,
      frontmatter,
    } as PageMeta;

    const result = `const frontmatter = ${JSON.stringify(frontmatter)};
${compileResult}
MDXContent.__RSPRESS_PAGE_META = {};

MDXContent.__RSPRESS_PAGE_META["${encodeURIComponent(
      normalizePath(path.relative(docDirectory, filepath)),
    )}"] = ${JSON.stringify(pageMeta)};
`;
    return result;
  } catch (e: unknown) {
    if (e instanceof Error) {
      throw e;
    }
    throw new Error(`MDX compile error: ${e} in ${filepath}`);
  }
}

const cache = new Map<string, Promise<string>>();

async function compileWithCrossCompilerCache(
  options: CompileOptions,
): Promise<string> {
  const filepath = options.filepath;
  const task = cache.get(filepath);
  if (task) {
    // only for web and node, one write and one read
    // free the memory after one read
    cache.delete(filepath);
    return task;
  }

  const promise = compile(options);
  cache.set(filepath, promise);
  return promise;
}

export { compile, compileWithCrossCompilerCache };
