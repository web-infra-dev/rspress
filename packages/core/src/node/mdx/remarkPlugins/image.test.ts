import { describe, it } from 'vitest';
import { compile } from '../processor';

describe('mdx', () => {
  it('basic', async () => {
    const result = await compile({
      source: `
![alt1](./test3.jpg)

<img src="./test4.png" alt="alt2" />
`,

      docDirectory: '/usr/rspress-project/docs',
      filepath: '/usr/rspress-project/docs/index.mdx',
      config: null,
      pluginDriver: null,
      routeService: null,
    });
    expect(result).toMatchSnapshot();
  });
});
