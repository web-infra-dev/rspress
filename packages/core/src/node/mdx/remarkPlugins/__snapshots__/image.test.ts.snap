// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mdx > basic 1`] = `
"const frontmatter = {};
import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import {useMDXComponents as _provideComponents} from "@mdx-js/react";
import image0 from "./test3.jpg";
import image1 from "./test4.png";
function _createMdxContent(props) {
  const _components = {
    img: "img",
    p: "p",
    ..._provideComponents(),
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.p, {
      children: _jsx(_components.img, {
        alt: "alt1",
        src: image0
      })
    }), "\\n", _jsx("img", {
      src: image1,
      alt: "alt2"
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = {
    ..._provideComponents(),
    ...props.components
  };
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}

MDXContent.__RSPRESS_PAGE_META = {};

MDXContent.__RSPRESS_PAGE_META["index.mdx"] = {"toc":[],"title":"","headingTitle":"","frontmatter":{}};
"
`;
